[postfix] Exclude password files

Iwao-Miyake opened the pull request:
Changed to sosreport exclude files set by following options.
- `lmtp_sasl_password_maps`
- `smtp_sasl_password_maps`
- `postscreen_dnsbl_reply_map`
- `smtp_sasl_auth_cache_name`

Resolvs : #3073

Signed-off-by: Iwao Miyake <miyake.iwao@fujitsu.com>

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [X] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [X] Is the subject and message clear and concise?
- [X] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [X] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [X] Are any related Issues or existing PRs [properly referenced](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) via a Closes (Issue) or Resolved (PR) line?

pmoravec commented on sos/report/plugins/postfix.py:
```diff
@@ -52,6 +52,31 @@ def forbidden_ssl_keys_files(self):
+                    if line.startswith('#'):
+                        continue
+                    words = line.split('=')
+                    if words[0].strip() in forbid_attributes:
+                        filepath = words[1].split()[0]
```
Assume a corner case:

    lmtp_sasl_password_maps = 

which would raise `IndexError: list index out of range` error ;-)

What about:

    words = line.replace(' ', '').split('=')
    if words[0] in forbid_attributes and len(words) > 1:
        filepath = words[1]

(this won't work if the file path contains space - is that syntax allowed?)

pmoravec commented:
It seems the commits stepon each others toes. Let do this on your local branch:

1) Copy the desired final `postfix.py` file somewhere (it will be updated during the process).
2) rebase your local branch:

    git rebase --interactive HEAD~4

3) Now, either:
- edit the file such that you `drop` all four commits, then copy the backed `postfix.py` to the proper location, write `git add sos/report/plugins/postfix.py` and `git commit -s` and type the commit message again
- or edit the rebase file by `drop`-ing 3 of 4 commits and `edit`-ing just one commit with the right text inside, then  copy the backed `postfix.py` to the proper location, write `git add sos/report/plugins/postfix.py` and `git commit --amend` (to amend the left commit)

4) `git rebase --continue` to replay on top of your branch head
5) `git push --force` your branch

Iwao-Miyake commented:
Thank you to guide me.
I did not understand about git.
So your guidance makes me helped.

I tried modifying the process to regex to solve the input no filepath.

pmoravec commented on sos/report/plugins/postfix.py:
```diff
@@ -52,6 +54,38 @@ def forbidden_ssl_keys_files(self):
+            with open(self.path_join('/etc/postfix/main.cf'), 'r') as cffile:
+                for line in cffile.readlines():
+                    # ignore comment and check option format
+                    line = re.sub('#.*', '', line)
```
It makes sense to scrub passwords also in comments, no? Two use cases:

1) Sometimes, previously used passwords are kept in comments as a kind of history. Even the historic passwords are worth to obfuscate, since they might be reused or still in use else-where (yes, neither that is a good password policy, but I have seen that in practise)

2) Sometimes, for debugging purposes, one can temporarily comment out existing password to test a password-less authentication.

Iwao-Miyake replied:
I think it may be a different issue from excluding files.
In my humble opinion, If necessary, we should create a new issue.

pmoravec replied:
Ah I apologize for my mental shortcut (we usually scrub passwords using regexp, while here we stop collecting files using regexp). I meant "stop collecting forbidden files mentioned also in comments". But I agree there would be less valid use cases for it, so I am OK with the current PR.

TurboTurtle commented on sos/report/plugins/postfix.py:
```diff
@@ -52,6 +54,38 @@ def forbidden_ssl_keys_files(self):
+                        if filepath.startswith(prefix):
+                            filepath = filepath[len(prefix):]
+                        fp.append(filepath)
+        finally:
+            return fp
```
A return in a `finally` block will swallow raised exceptions. I think we should instead have an exception handler (log it for traceability) and have the return outside the try/except block.

TurboTurtle reviewed (commented):
Overall looks good, just one comment below regarding the try block.

Iwao-Miyake commented:
I compiled the commit added logs, possibly is not this needed?

In forbidden_password_files(), I have added to check non-exists path.
And the following logging process have been added for traceability.
- warning non-exists path
- error when the process catch exception

and changed the 'return' position.

TurboTurtle commented on sos/report/plugins/postfix.py:
```diff
@@ -52,6 +55,51 @@ def forbidden_ssl_keys_files(self):
+                        # ignore non-existent files
+                        if not os.path.exists(filepath):
+                            # warning log for traceability
+                            msg = "'%s' has non-existent path '%s'" % \
+                                    (attribute, filepath)
+                            msg = "[plugin:%s] %s" % (self.name(), msg)
+                            self.soslog.warning(msg)
+                            continue
+                        fp.append(filepath)
+        except Exception as e:
+            # error log
+            msg = "catch '%s' in forbidden_password_files()" % e.args[0]
+            msg = "[plugin:%s] %s" % (self.name(), msg)
+            self.soslog.error(msg)
```
`_log_*()` methods handle proper formatting here.

E.G.

```python
self._log_info(f"Filepath {filepath} specified by {attribute} does not exist.")
```

or/and

```python
self._log_error(f"Error parsing main.cf: {e}")
```

pmoravec commented on sos/report/plugins/postfix.py:
```diff
@@ -52,6 +55,51 @@ def forbidden_ssl_keys_files(self):
+                        # remove prefix
+                        if filepath.startswith(prefix):
+                            filepath = filepath[len(prefix):]
+                        # ignore non-existent files
+                        if not os.path.exists(filepath):
+                            # warning log for traceability
+                            msg = "'%s' has non-existent path '%s'" % \
+                                    (attribute, filepath)
+                            msg = "[plugin:%s] %s" % (self.name(), msg)
+                            self.soslog.warning(msg)
+                            continue
```
This is not needed, imho. The worst what would happen is you feed `add_forbidden_path` with a non-existing file, and any file to be collected would be verified about bit bigger list (https://github.com/sosreport/sos/blob/91cce33a1c9b3cff0795c7e573813d3394cc4107/sos/report/plugins/__init__.py#L1413-L1416).
